crates/typst/src/layout/grid/lines.rs

```diff
@@ -0,0 +1,749 @@
+    I::IntoIter: 'grid,
+{
+    // The segment currently being drawn.
+    // It is extended for each consecutive track through which the line would
+    // be drawn with the same stroke.
+    // Starts as None to force us to create a new segment as soon as we find
+    // the first track through which we should draw.
+    let mut current_segment: Option<(Arc<Stroke<Abs>>, Abs, Abs)> = None;
+    // How far from the start (before the first track) have we gone so far.
+    // Used to determine the positions at which to draw each segment.
+    let mut offset = Abs::zero();
+    // How much to multiply line indices by to account for gutter.
+    let gutter_factor = if grid.has_gutter { 2 } else { 1 };
+    // Which line position to look for in the given list of lines.
+    // If the index represents a gutter track, this means the list of lines
+    // parameter will actually correspond to the list of lines in the previous
+    // index, so we must look for lines positioned after the previous index,
+    // and not before, to determine which lines should be placed in gutter.
+    // Note that the maximum index is always an odd number when there's gutter,
+    // so we must check for it to ensure we don't give it the same treatment as
+    // a line before a gutter track.
+    let expected_line_position = if grid.has_gutter && index % 2 == 1 && !is_max_index {
+        LinePosition::After
+    } else {
+        LinePosition::Before
+    };
+
+    // Create an iterator which will go through each track, from start to
+    // finish, to create line segments and extend them until they are
+    // interrupted. Each track will be mapped to the finished line segment
+    // they interrupted; if they didn't interrupt any, they are filtered out.
+    // When going through each track, we check if the current segment would be
+    // interrupted, either because, at this track, we hit a merged cell over
+    // which we shouldn't draw, or because the line would have a different
+    // stroke at this point (so we have to start a new segment). If so, the
+    // current segment is yielded and the variable is set to None (meaning we
+    // have to create a new one later) or to the new segment (if we're starting
+    // to draw a segment with a different stroke than before).
+    // Otherwise (if the current segment should span the current track), it is
+    // simply extended (or a new one is created, if it is 'None'), and no value
+    // is yielded for the current track, since the segment isn't yet complete
+    // (the next tracks might extend it further before it is interrupted and
+    // yielded). That is, we yield each segment only when it is interrupted,
+    // since then we will know its final length for sure.
+    // We chain an extra 'None' track to ensure the final segment is always
+    // interrupted and yielded, if it wasn't interrupted earlier.
+    tracks.into_iter().map(Some).chain(std::iter::once(None)).filter_map(
```

Comment:
This isn't a concrete suggestion, so feel free to ignore this, but I do wonder if this could be simplified and brougt closer to the original style it currently has on main (with vector) by using `std::iter::from_fn`.

Reply:
Done. It's a bit closer to the original style now, and perhaps more legible. Though it's a bit unfortunate that we lose the size hint and trait impls (such as `FusedIterator`) provided by `filter_map`, but I don't think we (read: `Vec::extend`) were really making use of those anyway (it uses the lower bound of the size hint, not the upper bound I believe, and the lower bound is 0 for the output of both `filter_map` and `std::iter::from_fn`).